crates/sui-core/src/execution_cache/writeback_cache.rs

```diff
@@ -0,0 +1,1158 @@
+
+        Ok(())
+    }
+
+    // Move an entry from the dirty queue to the cache queue. This is called after the entry is
+    // committed to the db.
+    fn move_version_from_dirty_to_cache<K, V>(
+        dirty: &DashMap<K, CachedVersionMap<V>>,
+        cache: &MokaCache<K, Arc<Mutex<CachedVersionMap<V>>>>,
+        key: K,
+        version: SequenceNumber,
+        value: &V,
+    ) where
+        K: Eq + std::hash::Hash + Clone + Send + Sync + Copy + 'static,
+        V: Send + Sync + Clone + Eq + std::fmt::Debug + 'static,
+    {
+        static MAX_VERSIONS: usize = 3;
+
+        // IMPORTANT: lock both the dirty set entry and the cache entry before modifying either.
+        // this ensures that readers cannot see a value temporarily disappear.
+        let cache_entry = cache.entry(key).or_default();
+        let mut cache_map = cache_entry.value().lock();
+        let dirty_entry = dirty.entry(key);
+
+        // insert into cache and drop old versions.
+        cache_map.insert(version, value.clone());
+        // TODO: make this automatic by giving CachedVersionMap an optional max capacity
+        cache_map.truncate(MAX_VERSIONS);
+
+        let DashMapEntry::Occupied(mut occupied_dirty_entry) = dirty_entry else {
+            panic!("dirty map must exist");
+        };
+
+        let removed = occupied_dirty_entry.get_mut().pop_oldest(&version);
+
+        assert_eq!(removed.as_ref(), Some(value), "dirty version must exist");
+
+        // if there are no versions remaining, remove the map entry
+        if occupied_dirty_entry.get().is_empty() {
+            occupied_dirty_entry.remove();
+        }
+    }
+
+    pub async fn prune_objects_and_compact_for_testing(
+        &self,
+        checkpoint_store: &Arc<CheckpointStore>,
+    ) {
+        let pruning_config = AuthorityStorePruningConfig {
+            num_epochs_to_retain: 0,
+            ..Default::default()
+        };
+        let _ = AuthorityStorePruner::prune_objects_for_eligible_epochs(
+            &self.store.perpetual_tables,
+            checkpoint_store,
+            &self.store.objects_lock_table,
+            pruning_config,
+            AuthorityStorePruningMetrics::new_for_test(),
+            usize::MAX,
+        )
+        .await;
+        let _ = AuthorityStorePruner::compact(&self.store.perpetual_tables);
+    }
+
+    pub fn store_for_testing(&self) -> &Arc<AuthorityStore> {
+        &self.store
+    }
+
+    pub fn as_notify_read_wrapper(self: Arc<Self>) -> NotifyReadWrapper<Self> {
+        NotifyReadWrapper(self)
+    }
+
+    #[cfg(test)]
+    pub fn clear_caches(&self) {
+        self.cached.object_cache.invalidate_all();
+        self.cached.packages.invalidate_all();
+        self.cached.marker_cache.invalidate_all();
+        self.cached._transaction_objects.invalidate_all();
+    }
+}
+
+impl ExecutionCacheAPI for WritebackCache {}
+
+impl ExecutionCacheCommit for WritebackCache {
+    fn commit_transaction_outputs(
```

Comment:
I know this PR doesn't use yet, but can you explain where we would be calling this function?

Reply:
added more documentation in the trait defn